lib/apps/timers.js

```diff
@@ -50,23 +50,110 @@ class BaseTimer {
+                startDateObj.setHours(null, null, null, 0);
+                let base_time = base - startDateObj;
+                interval /= this._frequency;
+                next_timing = Math.round(now + interval - ((now - base) % interval));
+                next_timing = new Date(next_timing);
+                next_timing.setHours(null, null, null, base_time);
+            } else {
+                // Calling more than once a day.
+                // User is being weird?
+                // Just do a simple divide
+                interval /= this._frequency;
+                next_timing = Math.round(now + interval - ((now - base) % interval));
+            }
+        }
+
+        // Just do simple divide if interval is less than a day
+        if (this._interval < DAY) {
+            if (base > now) {
+                next_timing = base;
+            } else {
+                interval /= this._frequency;
+                next_timing = Math.round(now + interval - ((now - base) % interval));
+            }
+        }
+
+        if (this._interval === DAY) {
+            let timings = [];
+            if (this._frequency === null || this._frequency === 1) {
+                // If it's just once a day, set timing as 12pm
+                timings.push(TIME_12PM); // 12pm
+            } else {
+                // For more than once a day, set interval as 9AM to 9PM
+                // and divide equally starting from 9AM onwards
+                // e.g. 3 times a day will be [9AM, 3PM, 9PM]
+                // 4 times a day will be [9AM, 1PM, 5PM, 9PM]
+                // Might want to set a limit on frequency?
+                // Like if it is more than once per hour
+                // then we just do simple divide
+                timings.push(TIME_9AM); // 9AM
+                interval = HALFDAY; // 9AM to 9PM
+                interval /= (this._frequency - 1);
+                for (var i = 0; i < (this._frequency - 1); i++) timings.push(Math.round(TIME_9AM + interval));
+            }
+
+            let threshold = Math.max(now, base);
+            next_timing = new Date(threshold);
+
+            // Set next timing as the earliest timing after threshold (base or now)
+            // e.g. if timings = [9am, 3pm, 9pm] and threshold = 1st Jan 1pm
+            // next timing should be 1st Jan 3pm
+            // TODO - there's a faster way than doing a while loop
+            let n = 0;
+            next_timing.setHours(null, null, null, timings[n]);
+            while (next_timing < threshold) {
+                if (n === timings.length - 1) {
+                    // Went past the latest timing, go to first timing of next day
+                    next_timing.setHours(null, null, null, timings[0]);
+                    next_timing = next_timing.valueOf() + DAY;
+                    break;
+                }
+                next_timing.setHours(null, null, null, timings[++n]);
+            }
+        }
+        return next_timing - now;
```

Comment:
Please move this to a separate function and add unit tests, so we know exactly what this code does.

Reply:
okay will do